charms/knative-operator/src/charm.py

```diff
@@ -0,0 +1,134 @@
+
+logger = logging.getLogger(__name__)
+
+
+class KnativeOperatorCharm(CharmBase):
+    """A Juju Charm for Training Operator"""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self.interfaces = ""
+
+        self._name = self.model.app.name
+        self._namespace = self.model.name
+        self._operator_service = "/ko-app/operator"
+        self.env = Environment(loader=FileSystemLoader('src'))
+        self._container = self.unit.get_container(self._name)
+        self._resources_files = {
+             "auth_manifests.yaml",
+             "service_account.yaml",
+             "crds_manifests.yaml",
+             # Skipping config manifests for now
+             # "config_manifests.yaml",
+        }
+        self._context = {"namespace": self._namespace, "name": self._name}
+
+        self.lightkube_client = Client(namespace=self._namespace, field_manager="lightkube")
+        for event in [self.on.install, self.on.config_changed]:
+            self.framework.observe(event, self._main)
+        self.framework.observe(
+            self.on.knative_operator_pebble_ready,
+            self._on_knative_operator_pebble_ready,
+        )
+
+    @property
+    def _knative_operator_layer(self) -> Layer:
+        """Returns a pre-configured Pebble layer."""
+
+        layer_config = {
+            "summary": "knative-operator layer",
+            "description": "pebble config layer for knative-operator",
+            "services": {
+                self._operator_service: {
+                    "override": "replace",
+                    "summary": "entrypoint of the knative-operator image",
+                    "command": self._operator_service,
+                    "startup": "enabled",
+                    "environment": {
+                        "POD_NAME": self._name,
+                        "SYSTEM_NAMESPACE": self._namespace,
+                        "METRICS_DOMAIN": "knative.dev/operator",
+                        "CONFIG_LOGGING_NAME": "config-loggig",
+                        "CONFIG_OBSERVABILITY_NAME": "config-observability",
+                    },
+                }
+            },
+        }
+        return Layer(layer_config)
+
+    def _update_layer(self, event) -> None:
+        """Updates the Pebble configuration layer if changed."""
+        if not self._container.can_connect():
+            self.unit.status = MaintenanceStatus("Waiting for pod startup to complete")
+            event.defer()
+            return
+
+        # Get current config
+        current_layer = self._container.get_plan()
+        # Create a new config layer
+        new_layer = self._knative_operator_layer
+        if current_layer.services != new_layer.services:
+            self._container.add_layer(self._operator_service, new_layer, combine=True)
+            logging.info("Pebble plan updated with new configuration")
+            self._container.restart(self._operator_service)
```

Comment:
This method restarts pebble, which (hopefully temporarily) disables the charm workload.  If `current_layer.services != new_layer.services`, should we be setting a `MaintenanceStatus`, doing the `.restart`, then in future (via defer or some wait here) checking if the charm is back online?

This feels like a general problem we need a pattern for rather than something specific to this charm, so if there's no easy fix we can ignore it for now and discuss separately to find a solution

Reply:
> should we be setting a MaintenanceStatus, doing the .restart, then in future (via defer or some wait here) checking if the charm is back online?

For the status, we are setting `MaintenanceStatus` in `_main`, not here since this is a helper method, I'm following more or less what we did for the training-operator, though I wouldn't mind moving that piece of code in this method.
As for checking if the service has been restarted correctly, `pebble` will throw some exceptions that we can handle, though it is true we have to have a standard way of doing that. I'll investigate and work on it.